sway/tree/container.c

```diff
@@ -109,10 +110,190 @@ static struct sway_container *_container_destroy(struct sway_container *cont) {
+		struct sway_container *new_workspace = NULL;
+		// TODO move floating
+		for (int i = 0; i < output->children->length; i++) {
+			if (output->children->items[i] != workspace) {
+				new_workspace = output->children->items[i];
+				break;
+			}
+		}
+
+		wlr_log(L_DEBUG, "moving children to different workspace '%s' -> '%s'",
+			workspace->name, new_workspace->name);
+		for (int i = 0; i < workspace->children->length; i++) {
+			container_move_to(workspace->children->items[i], new_workspace);
+		}
+	}
+
+	container_finish(workspace);
+	return parent;
+}
+
+
+static void reap_empty_func(struct sway_container *con, void *data) {
+	switch (con->type) {
+		case C_TYPES:
+		case C_ROOT:
+		case C_OUTPUT:
+			// dont reap these
+			break;
+		case C_WORKSPACE:
+			if (!workspace_is_visible(con) && con->children->length == 0) {
+				container_workspace_destroy(con);
+			}
+			break;
+		case C_CONTAINER:
+			if (con->children->length == 0) {
+				container_finish(con);
+			} else if (con->children->length == 1) {
+				struct sway_container *only_child = con->children->items[0];
+				if (only_child->type == C_CONTAINER) {
+					container_remove_child(only_child);
+					container_replace_child(con, only_child);
+					container_finish(con);
+				}
+			}
+		case C_VIEW:
+			break;
+	}
+}
+
+struct sway_container *container_reap_empty(struct sway_container *container) {
+	struct sway_container *parent = container->parent;
+
+	container_for_each_descendant_dfs(container, reap_empty_func, NULL);
+
+	return parent;
+}
+
+static void container_root_finish(struct sway_container *con) {
+	wlr_log(L_ERROR, "TODO: destroy the root container");
+}
+
+struct sway_container *container_destroy(struct sway_container *con) {
+	if (con == NULL) {
+		return NULL;
+	}
+
+	struct sway_container *anscestor = NULL;
+
+	switch (con->type) {
+		case C_ROOT:
+			container_root_finish(con);
+			break;
+		case C_OUTPUT:
+			anscestor = container_output_destroy(con);
+			break;
+		case C_WORKSPACE:
+			anscestor = container_workspace_destroy(con);
+			break;
+		case C_CONTAINER:
+			if (con->children != NULL && con->children->length) {
+				assert(false &&
+					"dont destroy container containers with children");
+			}
+			container_finish(con);
+			// TODO return parent to arrange maybe?
+			break;
+		case C_VIEW:
+			container_finish(con);
+			// TODO return parent to arrange maybe?
+			break;
+		case C_TYPES:
+			wlr_log(L_ERROR, "tried to destroy an invalid container");
+			break;
+	}
+
+	container_reap_empty(&root_container);
+
+	return anscestor;
+}
+
+static void container_close_func(struct sway_container *container, void *data) {
+	if (container->type == C_VIEW) {
+		view_close(container->sway_view);
+	}
+}
+
+struct sway_container *container_close(struct sway_container *con) {
+	if (!sway_assert(con != NULL,
+				"container_close called with a NULL container")) {
+		return NULL;
+	}
+
+	struct sway_container *parent = con->parent;
+
+	switch (con->type) {
+		case C_TYPES:
```

Comment:
In case of `C_TYPES` it should fail

Reply:
what's the reason for `C_TYPES`? It's never used except as an error condition.

Reply:
It might have been more useful before than it is today

Reply:
Yeah, I'd like to remove it.